qualtran/bloqs/mcmt/bloq_with_specialized_single_qubit_control.py

```diff
@@ -0,0 +1,102 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import cast, Iterable, Optional, Protocol, runtime_checkable, Sequence, TYPE_CHECKING
+
+import numpy as np
+
+if TYPE_CHECKING:
+    from qualtran import AddControlledT, Bloq, BloqBuilder, CtrlSpec, SoquetT
+
+
+@runtime_checkable
+class BloqWithSpecializedControl(Protocol):
+    """mixin for a bloq that has a specialized single qubit controlled version."""
+
+    @property
+    def cv(self) -> Optional[int]: ...
+
+    def with_cv(self, *, cv: Optional[int]) -> Optional['Bloq']: ...
+
+    @property
+    def ctrl_reg_name(self) -> str: ...
+
+
+def get_ctrl_system_for_bloq_with_specialized_single_qubit_control(
+    bloq: 'BloqWithSpecializedControl', ctrl_spec: 'CtrlSpec'
+) -> tuple['Bloq', 'AddControlledT']:
+    from qualtran import Bloq, CtrlSpec, Soquet
+    from qualtran.bloqs.mcmt import ControlledViaAnd
+
+    def _get_default_fallback():
+        assert isinstance(bloq, Bloq)
+        return ControlledViaAnd.make_ctrl_system(bloq=bloq, ctrl_spec=ctrl_spec)
+
+    if ctrl_spec.num_qubits != 1:
+        return _get_default_fallback()
+
+    assert isinstance(
+        bloq, BloqWithSpecializedControl
```

Comment:
so this function needs to get a bloq's current `cv`, it's control register name, and a callable that returns a controlled version.

Did you consider just taking these as arguments to the function? i.e. at the call site you could just pass a cv and control register name.

Maybe this is awkward for the `with_cv` callable. But do you expect this to basically always be `lambda cv: attrs.evolve(bloq, **{ctrl_attr_name: cv})` for a given `ctrl_attr_name`? if so: you can specialize to this case and take everything as arguments instead of a duck-typing `Protocol`.

Reply:
Yeah I think this is always the pattern. I realized it doesn't even need to be a callable, I just need 3 bloqs: `ctrl_bloq`, `unctrl_bloq`, `ctrl_0_bloq`. I can just pass those three bloqs here, along with the control register name and the current `cv`.